File: FMS.Simulator/Worker.vb
Imports FMS.Business
Imports System.Net

Public Class Worker

    Public Shared Sub DoWork()

        Console.WriteLine(vbNewLine & "FLEET MANAGEMENT ""DEMO"" SIMULATOR " & vbNewLine & vbNewLine & vbNewLine & vbNewLine)

        Dim settings As List(Of DataObjects.SimulatorSetting) = DataObjects.SimulatorSetting.GetAll

        Dim ts As TimeSpan = TimeSpan.FromSeconds(2)

        Dim timeiterators As New List(Of TimeIterator)


        For Each Setting As DataObjects.SimulatorSetting In settings

            Dim ti As New TimeIterator With {.settingObj = Setting}

            Dim sourceDevice As DataObjects.Device = DataObjects.Device.GetAllDevices.Where( _
                                                            Function(x) x.DeviceID = ti.settingObj.SourceDeviceID).Single

            Dim st As DateTime = ti.settingObj.StartTime
            Dim et As DateTime = ti.settingObj.EndTime

            While st < et

                Dim ll As New LatLongDate

                ll.datetime = st

                With sourceDevice.GetLatLongs(ll.datetime)
                    ll.latitude = .Key
                    ll.longitude = .Value
                End With

                ti.latlongs.Add(ll)

                st = st + ts

            End While

            timeiterators.Add(ti)

        Next


        Dim baseURL As String = "http://ppjs.nanosoft.com.au:9000/api/dataaccess?truckid={0}&lat={1}&lng={2}&time={3}"

        While True

            For Each x In timeiterators

                'using(WebClient client = new WebClient()) {
                'string s = client.DownloadString(url);
                '}

                Using client As New WebClient

                    Dim lld As LatLongDate = x.GetNext

                    Dim url As String = String.Format(baseURL, x.settingObj.DestinationDeviceID, lld.latitude, lld.longitude, Now.ToString("dd/MMM/yyyy HH:mm:ss"))

                    Dim s As String = client.DownloadString(url)
                End Using


            Next

            Threading.Thread.Sleep(ts)
        End While

    End Sub


End Class


''' <summary>
''' Starts at the beginning of the list and iterates through to the top 
''' THEN, goes from the top to the bottom again, repeates indefinitley
''' </summary>
Public Class TimeIterator

    Public Property settingObj As DataObjects.SimulatorSetting

    Public Property latlongs As New List(Of LatLongDate)

    ''' <summary>
    ''' The direction which the list iterates, this should be set to either +1 or -1
    ''' </summary>
    Public Property direction As Integer = 1
    Public Property currentIndex As Integer = 0

    Public Function GetNext() As LatLongDate

        Dim maxIndex As Integer = latlongs.Count - 1

        If currentIndex = 0 Then direction = 1
        If currentIndex = maxIndex Then direction = -1

        currentIndex = currentIndex + direction

        'get the  next index
        Return latlongs(currentIndex)

    End Function

End Class


Public Class LatLongDate

    Public Property latitude As Decimal
    Public Property longitude As Decimal
    Public Property datetime As DateTime

    Public Sub New()

    End Sub

End Class

File: FMS.SimulatorService/Classes/Worker.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using FMS.Business;
using System.Net;

namespace FMS.SimulatorService.Classes
{
    public class Worker
    {

        private static Topshelf.Logging.LogWriter logger = Topshelf.Logging.HostLogger.Get<Worker>();

        public static void DoWork()
        {


            logger.Info( "\nFLEET MANAGEMENT \"DEMO\" SIMULATOR \n\n\n");


            List<FMS.Business.DataObjects.SimulatorSetting> settings = FMS.Business.DataObjects.SimulatorSetting.GetAll();

            TimeSpan ts = TimeSpan.FromSeconds(2);

            List<TimeIterator> timeiterators = new List<TimeIterator>();

            

            foreach (FMS.Business.DataObjects.SimulatorSetting Setting in settings)
            {


                logger.Info(Setting);

                TimeIterator ti = new TimeIterator { settingObj = Setting };

                FMS.Business.DataObjects.Device sourceDevice = (from x in FMS.Business.DataObjects.Device.GetAllDevices()
                                                                where x.DeviceID == ti.settingObj.SourceDeviceID
                                                                select x).Single();


                DateTime st = ti.settingObj.StartTime;
                DateTime et = ti.settingObj.EndTime;


                while (st < et)
                {
                    LatLongDate ll = new LatLongDate();

                    ll.datetime = st;

                    var _with1 = sourceDevice.GetLatLongs(ll.datetime);
                    ll.latitude = _with1.Key;
                    ll.longitude = _with1.Value;

                    ti.latlongs.Add(ll);
                    st = st + ts;

                }

                timeiterators.Add(ti);

            }


            string baseURL = @"http://{0}.nanosoft.com.au:9000/api/dataaccess?truckid={1}&lat={2}&lng={3}&time={4}";


            while (true)
            {

                foreach (var x_loopVariable in timeiterators)
                {

                    var x = x_loopVariable;
                    //using(WebClient client = new WebClient()) {
                    //string s = client.DownloadString(url);
                    //}

                    using (WebClient client = new WebClient())
                    {

                        LatLongDate lld = x.GetNext();

                        string url = string.Format(baseURL, Properties.Settings.Default.destination_server
                                                            , x.settingObj.DestinationDeviceID
                                                            , lld.latitude
                                                            , lld.longitude
                                                            , DateTime.Now.ToString("dd/MMM/yyyy HH:mm:ss"));

                        logger.Debug(url);

                        string s = client.DownloadString(url);
                    }


                }

                System.Threading.Thread.Sleep(ts);
            }

        }


    }


    /// <summary>
    /// Starts at the beginning of the list and iterates through to the top 
    /// THEN, goes from the top to the bottom again, repeates indefinitley
    /// </summary>
    public class TimeIterator
    {

        public FMS.Business.DataObjects.SimulatorSetting settingObj { get; set; }

        public List<LatLongDate> latlongs { get; set; }

        public TimeIterator()
        {

            latlongs = new List<LatLongDate>();
        }

        /// <summary>
        /// The direction which the list iterates, this should be set to either +1 or -1
        /// </summary>
        public int direction { get; set; }
        public int currentIndex { get; set; }

        public LatLongDate GetNext()
        {

            int maxIndex = latlongs.Count - 1;

            if (currentIndex == 0)
                direction = 1;
            if (currentIndex == maxIndex)
                direction = -1;

            currentIndex = currentIndex + direction;

            //get the  next index
            return latlongs[currentIndex];

        }

    }


    public class LatLongDate
    {

        public decimal latitude { get; set; }
        public decimal longitude { get; set; }
        public DateTime datetime { get; set; }


        public LatLongDate()
        {
        }

    }
}